pkg/flowaggregator/clickhouseclient/clickhouseclient.go

```diff
@@ -487,24 +496,33 @@ func (ch *ClickHouseExportProcess) batchCommitAll() (int, error) {
 
 		if err != nil {
 			klog.ErrorS(err, "Error when adding record")
+			ch.pushFrontRecords(recordsToExport)
 			_ = tx.Rollback()
 			return 0, err
 		}
 	}
 
 	if err := tx.Commit(); err != nil {
 		klog.ErrorS(err, "Error when committing record")
+		ch.pushFrontRecords(recordsToExport)
 		return 0, err
 	}
 
-	// remove committed record from deque
+	return len(recordsToExport), nil
+}
+
+// pushFrontRecords pushes records back to the front of deque without exceeding its capacity.
+// Items with lower index (older records) will be dropped first if deque is to be filled.
+func (ch *ClickHouseExportProcess) pushFrontRecords(records []*ClickHouseFlowRow) {
```

Comment:
let's name this `pushRecordsToFrontOfQueue`

Reply:
done